R/generics.R

```diff
@@ -102,8 +102,28 @@ dplyr_col_modify <- function(data, cols) {
 
 #' @export
 dplyr_col_modify.data.frame <- function(data, cols) {
-  data[names(cols)] <- cols
-  data
+  # Implement from first principles to avoiding edge cases in [.data.frame
+  # and [.tibble in 2.1.3 and earlier
+
+  # Apply tidyverse recycling rules
+  cols <- lapply(cols, vec_recycle, size = nrow(data))
+
+  names(cols) <- as_utf8_character(names2(cols))
+  names(data) <- as_utf8_character(names2(data))
+
+  out <- vec_data(data)
+  for (i in seq_along(cols)) {
+    nm <- names(cols)[[i]]
+    out[[nm]] <- cols[[i]]
+  }
```

Comment:
```suggestion
  out <- vec_data(data)

  nms <- as_utf8_character(names2(cols))
  names(out) <- as_utf8_character(names2(out))

  for (i in seq_along(cols)) {
    nm <- nms[[i]]
    out[[nm]] <- cols[[i]]
  }
```

Mainly this avoids the repeated call to `names(cols)[[i]]` in the loop, but also calls `vec_data()` earlier to avoid data frame dispatch on `names()` and `names<-()`.